service/api_task.go

```diff
@@ -635,6 +634,26 @@ func handleOldAgentRevision(response apimodels.NextTaskResponse, h *host.Host, w
 			gimlet.WriteJSON(w, response)
 			return apimodels.NextTaskResponse{}, true
 		}
+
+		// Non-legacy hosts deploying agents via the agent monitor may be
+		// running an agent on the current revision, but the database host has
+		// yet to be updated.
+		if !h.LegacyBootstrap() && details.AgentRevision != h.AgentRevision {
+			if err := h.SetAgentRevision(details.AgentRevision); err != nil {
+				grip.Error(message.WrapError(err, message.Fields{
+					"message":       "problem updating host agent revision",
+					"operation":     "next_task",
+					"host":          h.Id,
+					"source":        "database error",
+					"host_revision": details.AgentRevision,
+					"revsision":     evergreen.BuildRevision,
+				}))
+			} else {
```

Comment:
I don't follow what's happening in this else. When will it be the case that the agent revision is old, but `details.AgentRevision == h.AgentRevision`, and why is that when we should log host agent deployed?

Reply:
Nested ifs are hard - this is the else case for the `SetAgentRevision()` call.

The app server should log to the event log in the case that `SetAgentRevision` succeeds (i.e. its error is `nil`). If there's a databse error and we can't update the db host's agent revision, it shouldn't log that a new agent has been deployed. Otherwise, the next time it hits this block of code, it will write a duplicate event log that the agent was deployed.

I rewrote this block to be easier to understand the logic.